src/primitives/gf32.rs

```diff
@@ -0,0 +1,463 @@
+    pub const D: Fe32 = Fe32(13);
+    /// Numeric value maps to bech32 character: 14 == "w".
+    pub const W: Fe32 = Fe32(14);
+    /// Numeric value maps to bech32 character: 15 == "0".
+    pub const _0: Fe32 = Fe32(15);
+    /// Numeric value maps to bech32 character: 16 == "s".
+    pub const S: Fe32 = Fe32(16);
+    /// Numeric value maps to bech32 character: 17 == "3".
+    pub const _3: Fe32 = Fe32(17);
+    /// Numeric value maps to bech32 character: 18 == "j".
+    pub const J: Fe32 = Fe32(18);
+    /// Numeric value maps to bech32 character: 19 == "n".
+    pub const N: Fe32 = Fe32(19);
+    /// Numeric value maps to bech32 character: 20 == "5".
+    pub const _5: Fe32 = Fe32(20);
+    /// Numeric value maps to bech32 character: 21 == "4".
+    pub const _4: Fe32 = Fe32(21);
+    /// Numeric value maps to bech32 character: 22 == "k".
+    pub const K: Fe32 = Fe32(22);
+    /// Numeric value maps to bech32 character: 23 == "h".
+    pub const H: Fe32 = Fe32(23);
+    /// Numeric value maps to bech32 character: 24 == "c".
+    pub const C: Fe32 = Fe32(24);
+    /// Numeric value maps to bech32 character: 25 == "e".
+    pub const E: Fe32 = Fe32(25);
+    /// Numeric value maps to bech32 character: 26 == "6".
+    pub const _6: Fe32 = Fe32(26);
+    /// Numeric value maps to bech32 character: 27 == "m".
+    pub const M: Fe32 = Fe32(27);
+    /// Numeric value maps to bech32 character: 28 == "u".
+    pub const U: Fe32 = Fe32(28);
+    /// Numeric value maps to bech32 character: 29 == "a".
+    pub const A: Fe32 = Fe32(29);
+    /// Numeric value maps to bech32 character: 30 == "7".
+    pub const _7: Fe32 = Fe32(30);
+    /// Numeric value maps to bech32 character: 31 == "l".
+    pub const L: Fe32 = Fe32(31);
+
+    /// Iterator over all field elements, in alphabetical order.
+    pub fn iter_alpha() -> impl Iterator<Item = Fe32> {
+        [
+            Fe32::A,
+            Fe32::C,
+            Fe32::D,
+            Fe32::E,
+            Fe32::F,
+            Fe32::G,
+            Fe32::H,
+            Fe32::J,
+            Fe32::K,
+            Fe32::L,
+            Fe32::M,
+            Fe32::N,
+            Fe32::P,
+            Fe32::Q,
+            Fe32::R,
+            Fe32::S,
+            Fe32::T,
+            Fe32::U,
+            Fe32::V,
+            Fe32::W,
+            Fe32::X,
+            Fe32::Y,
+            Fe32::Z,
+            Fe32::_0,
+            Fe32::_2,
+            Fe32::_3,
+            Fe32::_4,
+            Fe32::_5,
+            Fe32::_6,
+            Fe32::_7,
+            Fe32::_8,
+            Fe32::_9,
+        ]
+        .iter()
+        .copied()
+    }
+
+    /// Creates a field element from a single bech32 character.
+    pub fn from_char(c: char) -> Result<Fe32, Error> {
+        let byte = i8::try_from(u32::from(c)).map_err(|_| Error::InvalidChar(c))?;
+        let byte = byte as u8; // cast guaranteed to be ok since we started with an unsigned value
```

Comment:
Why not cast to `usize` here instead of below?

Reply:
Good call. I think it we just iterated too many times and wound up with goofy code. Agreed, casting to `usize` directly would be better.

Reply:
Changing the cast made me read this code again and think a bit harder. I then simplified it to use explicit value checks instead of using the type conversions to implicitly ensure the array access is within bounds. Next I added a kani verification and fixed the bugs I'd introduced - profit!